src/napari_builtins/_qt/features_table.py

```diff
@@ -430,90 +456,185 @@ def _get_selection_event_for_layer(layer):
+        # Get all selected row indices from the table
+        selected_global_rows = [
+            self.table.model().mapToSource(raw_index).row()
+            for raw_index in self.table.selectionModel().selectedIndexes()
+        ]
+
+        if not selected_global_rows:
+            return
+
+        df = self.table.model().sourceModel().df
+        if df.empty:
+            return
+
+        # Calculate layer start indices for all layers once (most efficient)
+        layer_starts = df.groupby('layer', sort=False, observed=False).apply(
+            lambda x: x.index[0], include_groups=False
+        )
+
+        # Get layer names for selected rows and convert to layer-specific indices
+        selected_layer_names = df['layer'].iloc[selected_global_rows]
+        layer_start_indices = selected_layer_names.map(layer_starts).astype(
+            int
+        )
+        layer_specific_indices = (
+            np.array(selected_global_rows) - layer_start_indices.values
+        )
+
+        # Group by layer name efficiently
+        selections_by_layer = {}
+        for layer_name, layer_idx in zip(
+            selected_layer_names, layer_specific_indices, strict=False
+        ):
+            selections_by_layer.setdefault(layer_name, []).append(layer_idx)
+
+        # Update layer selections
+        for layer in self._selected_layers:
+            if layer.name not in selections_by_layer:
+                continue
+            layer_indices = selections_by_layer[layer.name]
 
-        elif hasattr(self._active_layer, 'selected_data'):
-            selected_rows = {
-                self.table.model().mapToSource(raw_index).row()
-                for raw_index in self.table.selectionModel().selectedIndexes()
-            }
             with self._block_selection():
-                self._active_layer.selected_data = selected_rows
+                if hasattr(layer, 'selected_label'):
+                    layer.selected_label = layer_indices[-1]
+                elif hasattr(layer, 'selected_data'):
+                    layer.selected_data = set(layer_indices)
 
-    def _on_layer_selection_changed(self):
+    def _update_table_selected_cells(self):
         """Update table selected cells when layer selection changes."""
         if self._selection_blocked:
             return
 
         model = self.table.model()
-        if hasattr(self._active_layer, 'selected_label'):
-            sel = self._active_layer.selected_label
-            indices = [sel]
-        elif hasattr(self._active_layer, 'selected_data'):
-            sel = self._active_layer.selected_data
-            indices = sel
+        df = model.sourceModel().df
+        if df.empty:
+            return
 
+        indices = []
+        for layer in self._selected_layers:
+            if hasattr(layer, 'selected_label'):
+                sel = layer.selected_label
+                layer_data_row_index = df[df['layer'] == layer.name].index
```

Comment:
Maybe identifying layers by their `layer.uuid` is a more secure way of attributing values to layers?

Reply:
> and they are also unique, right?

Should be, yes.